crates/ide/src/syntax_highlighting/highlight.rs

```diff
@@ -226,35 +226,16 @@ pub(super) fn element(
                 T![unsafe] => h | HlMod::Unsafe,
                 T![true] | T![false] => HlTag::BoolLiteral.into(),
                 T![self] => {
-                    let self_param_is_mut = element
-                        .parent()
-                        .and_then(ast::SelfParam::cast)
-                        .and_then(|p| p.mut_token())
-                        .is_some();
-                    let self_path = &element
-                        .parent()
-                        .as_ref()
-                        .and_then(SyntaxNode::parent)
-                        .and_then(ast::Path::cast)
-                        .and_then(|p| sema.resolve_path(&p));
-                    let mut h = HlTag::Symbol(SymbolKind::SelfParam).into();
-                    if self_param_is_mut
-                        || matches!(self_path,
-                            Some(hir::PathResolution::Local(local))
-                                if local.is_self(db)
-                                    && (local.is_mut(db) || local.ty(db).is_mutable_reference())
-                        )
+                    let self_param = element.parent().and_then(ast::SelfParam::cast);
+                    if let Some(NameClass::Definition(def)) = self_param
+                        .and_then(|self_param| NameClass::classify_self_param(sema, &self_param))
                     {
-                        h |= HlMod::Mutable
-                    }
-
-                    if let Some(hir::PathResolution::Local(local)) = self_path {
-                        if is_consumed_lvalue(element, &local, db) {
-                            h |= HlMod::Consuming;
-                        }
+                        highlight_def(db, def) | HlMod::Definition
+                    } else if element.ancestors().any(|it| it.kind() == USE_TREE) {
+                        HlTag::Symbol(SymbolKind::SelfParam).into()
+                    } else {
+                        return None;
```

Comment:
I'm not sure if there is a better way to prevent this but if we don't return None here we overwrite the highlight modifiers that are applied when checking the `NameRef` for a self 😕

Reply:
I think its fine. But I am thinking: if `self` is always a Name or a NameRet, perhaps we could just

```
T![self] => return None,
```

?

Reply:
It's only a NameRef when in a path, so a SelfParam still has the keyword in it directly without a wrapping `Name`, hence the `classify_self_param`, but ye otherwise it should always take the NameRef arm. The `UseTree` check is there cause it would otherwise end up as unresolved reference in the NameRef arm, so it would need an extra check there at least instead.

Reply:
I mean, shouldn't it be wrapped in a name?

Reply:
Oh now I see, that should almost work out then, there are still two places that refer to the self keyword though. ExternCrate, which I think can just be replaced as well as its defined as:
```
ExternCrate =
  Attr* Visibility?
  'extern' 'crate' (NameRef | 'self') Rename? ';'
```
and `Visibility`:
```
Visibility =
  'pub' ('('
    'super'
  | 'self'
  | 'crate'
  | 'in' Path
  ')')?
```
Should we replace this as well with `NameRef | 'in' Path` then?

Reply:
Good question! 

looking at this, shouldn't this be just `'in'? Path`? That is, parse short forms as paths. 

extern crate should be just nameret now

Reply:
Right, it being `'in'? Path` makes more sense. Guess I'll go work on that as well while I'm already here.